### PR TITLE
Robert/189 remove abstract facade test

### PR DESCRIPTION
**Pull request by RobertSaylesAND**
[Ticket](https://github.com/isaaccomputerscience/isaac-cs-issues/issues/189)
Reimplementing test cases from `QuestionFacadeTest` and `QuizFacadeTest` in newer integration test framework to allow for removal of deprecated `AbstractFacadeTest` class.

Please note that, per recommendation, the success-case test for the `answerQuestion` endpoint has been extracted to its own [ticket](https://github.com/isaaccomputerscience/isaac-cs-issues/issues/226) due to issues with the question validation modules/injection under the IT framework that need to be resolved.

### CODE REVIEW

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/IsaacIntegrationTest.java [hunk 1]
```diff
@@ -136,6 +136,7 @@ public abstract class IsaacIntegrationTest {
   protected static RecaptchaManager recaptchaManager;
   protected static PgUsers pgUsers;
   protected static PgAnonymousUsers pgAnonymousUsers;
+  protected static ContentMapper contentMapper;
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/IsaacIntegrationTest.java:139]: Small one, but this and the two lines above are in a block of code called "Managers", so they probably belong somewhere else.

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java [hunk 2]
```diff
@@ -70,6 +92,38 @@ public final class ITConstants {
...
+
+  public static final Long QUIZ_ASSIGNMENT_ID = 1L;
+  public static final Long QUIZ_ASSIGNMENT_FOR_CANCELLATION_TEST_FIRST_ID = 2L;
+  public static final Long QUIZ_ASSIGNMENT_FOR_CANCELLATION_TEST_SECOND_ID = 3L;
+  public static final Long QUIZ_ASSIGNMENT_SECOND_ID = 4L;
+  public static final Long QUIZ_ASSIGNMENT_CANCELLED_ID = 5L;
+  public static final Long QUIZ_ASSIGNMENT_EXPIRED_ID = 6L;
+  public static final Long QUIZ_ASSIGNMENT_FEEDBACK_MODE_ID = 7L;
+  public static final Long QUIZ_ASSIGNMENT_SET_INCOMPLETE_TEST_ID = 8L;
+  public static final long QUIZ_ASSIGNMENT_NON_EXISTENT_ID = 1000L;
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java:108]: Another really small one, but shouldn't this be `Long`, just for consistency? There are also three spaces after this. Does that mean that a comment is missing there?

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java [hunk 3]
```diff
@@ -64,202 +120,3415 @@ public class QuizFacadeIT extends IsaacIntegrationTest {
...
+    }
+
+    @Test
+    public void getQuizzesAsStudent_returnsNonInvisibleToStudentQuizzes()
+        throws NoCredentialsAvailableException, NoUserException, SegueDatabaseException,
+        AuthenticationProviderMappingException, IncorrectCredentialsProvidedException,
+        AdditionalAuthenticationRequiredException, InvalidKeySpecException, NoSuchAlgorithmException,
+        MFARequiredButNotConfiguredException {
+      // Arrange
+      // log in as Tutor, create request
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:218]: It looks like this is logging in as a student, not a tutor.

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java [hunk 4]
```diff
@@ -36,15 +38,32 @@ public final class ITConstants {
   public static final long TEST_TEACHER_ID = 5L;
 
   public static final String DAVE_TEACHER_EMAIL = "dave-teacher@test.com";
+  public static final String DAVE_TEACHER_PASSWORD = "test1234";
 
   public static final String TEST_STUDENT_EMAIL = "test-student@test.com";
   public static final String TEST_STUDENT_PASSWORD = "test1234";
   public static final long TEST_STUDENT_ID = 6L;
 
+  public static final String TEST_STUDENT_ALICE_EMAIL = "alice-student@test.com";
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java:47]: What's the reason to have these new test users' names as part of the variable? What makes Alice, Bob, Erika and Charlie different to the test student with ID 6? Are they just additional students? If so, you may want to name them student 2, student 3, etc. Just because you already have a test student, so it looks strange to have one unnamed student and then all these named students. If, on the other hand, there is something special about these new users, then that may be a sign that the name should reflect that.

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java [hunk 5]
```diff
@@ -64,202 +120,3415 @@ public class QuizFacadeIT extends IsaacIntegrationTest {
...
+        HttpServletRequest getQuizAssignmentRequest = createRequestWithCookies(new Cookie[] {teacherLogin.cookie});
+        replay(getQuizAssignmentRequest);
+
+        Response getQuizAssignmentResponse =
+            quizFacade.getQuizAssignment(getQuizAssignmentRequest, QUIZ_ASSIGNMENT_CANCELLED_ID);
+
+        assertEquals(Response.Status.BAD_REQUEST.getStatusCode(), getQuizAssignmentResponse.getStatus());
+
+        assertEquals("This assignment has been cancelled.",
+            getQuizAssignmentResponse.readEntity(SegueErrorResponse.class).getErrorMessage());
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:420]: Having all these chained calls makes your code hard to read. In cases like this, it's more readable to simply save the result in a variable. For instance, I would have written this assertion like this:
```java
String expectedErrorMessage = "This assignment has been cancelled.";
String actualErrorMessage = getQuizAssignmentResponse.readEntity(SegueErrorResponse.class).getErrorMessage();
assertEquals(expectedErrorMessage, actualErrorMessage);
```

Or even like this:
```java
SegueErrorResponse errorResponse = getQuizAssignmentResponse.readEntity(SegueErrorResponse.class);
String actualErrorMessage = errorResponse.getErrorMessage();
String expectedErrorMessage = "This assignment has been cancelled.";
assertEquals(expectedErrorMessage, actualErrorMessage);
```

That does not mean that you have to do this for every single assertion. When you have short assertions like the following, it's fine, because they're easily readable:
```java
assertEquals(QUIZ_TEST_QUIZ_ID, responseBody.getQuizId());
```

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java [hunk 6]
```diff
@@ -64,202 +120,3415 @@ public class QuizFacadeIT extends IsaacIntegrationTest {
...
-    SegueErrorResponse responseBody = (SegueErrorResponse) previewQuizResponse.getEntity();
-    assertEquals("You do not have the permissions to complete this action", responseBody.getErrorMessage());
+  /**
+   * As the integration tests do not currently support MFA login, we cannot use the normal login process and have to
+   * create cookies manually when testing admin accounts.
+   *
+   * @return a Cookie loaded with session information for the test admin user.
+   * @throws JsonProcessingException if the cookie serialisation fails
+   */
+  public Cookie createManualCookieForAdmin() throws JsonProcessingException {
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:3520]: Shouldn't this be private?

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java [hunk 7]
```diff
@@ -18,34 +18,90 @@
...
+import static uk.ac.cam.cl.dtg.segue.api.Constants.NUMBER_SECONDS_IN_ONE_DAY;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
 import jakarta.servlet.http.Cookie;
 import jakarta.servlet.http.HttpServletRequest;
 import jakarta.ws.rs.core.Request;
 import jakarta.ws.rs.core.Response;
 import java.security.NoSuchAlgorithmException;
 import java.security.spec.InvalidKeySpecException;
+import java.text.SimpleDateFormat;
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:91]: The following three imports belong to the old time API, and we shouldn't be using them:
```java
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
```

There are more modern alternatives. Here's some info:
https://programminghints.com/2017/05/still-using-java-util-date-dont/
https://www.baeldung.com/migrating-to-java-8-date-time-api

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/IsaacIntegrationTest.java (hunk 1 above)
Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/IsaacIntegrationTest.java:139]: Oh, possibly. I overlooked the header and just put it roughly in the order it was created. I'll have a look and figure out a better place for it...

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java (hunk 3 above)
Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:218]: Iirc, that block of tests is the only one that existed before the ticket and wasn't replaced with a newer version. It looked to cover all the cases it needed to but I guess I'll need to give it another look over...

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java (hunk 4 above)
Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java:47]: Those are the pre-existing names used in the test database, I just followed the naming pattern established by the other entries.
I don't know why they used the generic alphabet names rather than numbers but I can have a look into changing it if you'd prefer?

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java (hunk 7 above)
Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:91]: I'm using Date because that's the type the endpoints accept. Changing the endpoints to take a different type is beyond the scope of this ticket and without that using a different type here would just be introducing unnecessary conversions.

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java (hunk 6 above)
Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:3520]: I did consider moving this over to the base test class alongside the existing loginAs method for if we need to test other classes with an admin. It's not being used anywhere else _yet_ though, so leaving it in here as private for now could also make sense... Which would you recommend?

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java (hunk 4 above)
Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java:47]: My only complain about this is that I cannot tell just by looking at this why are they needed. Why cannot we just use the existing`TEST_STUDENT_EMAIL`? If there is a special reason why we need these new ones, then these constant names should reflect that reason, so that it's evident why they're here. For instance, `TEST_STUDENT_EMAIL` is pretty clear: it's the email of a test student. But `TEST_STUDENT_ALICE_EMAIL` is not a great name because it does not tell me why a student named Alice is needed, if that makes sense.

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java (hunk 6 above)
Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java]: Unless I know for sure that I'm going to need something in the near future (for instance, because there is a ticket in the current or next sprint that we're definitely going to be working on), I never do things just in case we need them in the future. If you google YAGNI and KISS principles, it will explain it better than I can.

TL;DR: `private`.

Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java]: This function should be moved back to `QuizFacadeIT`, again because of YAGNI and KISS principles. You've moved it to `IsaacIntegrationTest`, but it's only used here, so it's here where it belongs. If at some point you need it in another test, then yes, you can move it to `IsaacIntegrationTest`, or even better, somewhere else and use composition, instead of inheritance, to get that functionality. Does this all make sense? I'm happy to explain further if it doesn't.

Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java]: Understood, I'll move that back and set it to private for now 👍

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java (hunk 4 above)
Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java:47]: Fair enough, I'll take another look at the constant names and see if I can distill it down to something more explanatory...

Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java]: Okay, I've changed the constants a bit to try an improve the clarity. Also added some supplementary explanation in comments, since it would be either unclear or unwieldy to try and put the whole thing in the constant.
Let me know if you have any suggestions to improve it further.

Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java]: Thank you very much for changing this, I think it is easier to understand now.
I think that the names would make more sense if they read like the following: for instance, instead of `QUIZ_FACADE_TEST_STUDENT_3_EMAIL_WITH_FREE_ATTEMPT` it should be `QUIZ_FACADE_TEST_STUDENT_3_WITH_FREE_ATTEMPT_EMAIL`. What do you think?

Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/ITConstants.java]: I wasn't sure which way round to put those bits myself, so if you think that would be better, I'm happy to swap them over. 👍

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java [hunk 8]
```diff
@@ -64,202 +122,3533 @@ public class QuizFacadeIT extends IsaacIntegrationTest {
...
+
+    @Nested
+    class CorrectGroupOwnerButInvalidTarget {
+      @Test
+      public void asTeacher_withTargetStudent_whoIsNotInGroup_isForbidden()
+          throws NoCredentialsAvailableException, NoUserException, SegueDatabaseException,
+          AuthenticationProviderMappingException, IncorrectCredentialsProvidedException,
+          AdditionalAuthenticationRequiredException, InvalidKeySpecException, NoSuchAlgorithmException,
+          MFARequiredButNotConfiguredException {
+        LoginResult teacherLogin = loginAs(httpSession, TEST_TEACHER_EMAIL, TEST_TEACHER_PASSWORD);
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:695]: Another way you can reduce the size of this file and make it easier to read is by moving stuff that you keep repeating in multiple tests, like these three lines, into their own function. For instance, these three lines could be a function called `prepareTeacherRequest` (or something like that), which returns the `HttpServletRequest`, already replayed.

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java [hunk 9]
```diff
@@ -64,202 +122,3533 @@ public class QuizFacadeIT extends IsaacIntegrationTest {
...
+      HttpServletRequest getQuizzesRequest = createRequestWithCookies(new Cookie[] {teacherLogin.cookie});
+      replay(getQuizzesRequest);
+
+      Response getQuizzesResponse = quizFacade.getAvailableQuizzes(createNiceMock(Request.class), getQuizzesRequest);
+
+      assertEquals(Response.Status.OK.getStatusCode(), getQuizzesResponse.getStatus());
+
+      @SuppressWarnings("unchecked") ResultsWrapper<QuizSummaryDTO> responseBody =
+          (ResultsWrapper<QuizSummaryDTO>) getQuizzesResponse.getEntity();
+      boolean isTestQuizPresent = responseBody.getResults().stream().anyMatch(q -> q.getId().equals(QUIZ_TEST_QUIZ_ID));
```
[the remark this answers is not in the document]
Comment (RobertSaylesAND) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:166]: Looks like that method is only in that class actually, the ones in other classes are similar but check for different getters. Not sure if I can generalise that step...

**Review by jmgq — APPROVED** [review]
